umbra-js/test/Umbra.test.ts

```diff
@@ -176,172 +206,248 @@ describe('Umbra class', () => {
-        r,
-        s
-      );
-      const expectedAmountReceived = BigNumber.from(quantity).sub(sponsorFee);
-      verifyEqualValues(await dai.balanceOf(destinationWallet.address), expectedAmountReceived);
-      verifyEqualValues(await dai.balanceOf(stealthKeyPair.address), 0);
-      verifyEqualValues(await dai.balanceOf(sponsorWallet.address), sponsorFee);
+  const sendTests = [
+    { id: 'send', name: 'Send' },
+    { id: 'batchSend', name: 'Batch send' },
+  ];
+
+  for (const test of sendTests) {
+    describe(`${test.name}, scan, and withdraw funds`, () => {
+      beforeEach(() => {
+        // Seems we somehow lose the provider attached to our sender, so make sure it's there. Without this
+        // some tests below throw with "Error: missing provider (operation="sendTransaction", code=UNSUPPORTED_OPERATION, version=abstract-signer/5.0.12)"
+        sender = sender.connect(ethers.provider);
+      });
+
+      const mintAndApproveDai = async (signer: Wallet, user: string, amount: BigNumber) => {
+        await dai.connect(signer).mint(user, amount);
+        await dai.connect(signer).approve(umbra.umbraContract.address, ethers.constants.MaxUint256);
+        await dai.connect(signer).approve(umbra.batchSendContract.address, ethers.constants.MaxUint256);
+      };
+
+      it('reverts if sender does not have enough tokens', async () => {
+        const msg = `Insufficient balance to complete transfer. Has 0 tokens, tried to send ${quantity.toString()} tokens.`;
+
+        if (test.id === 'send') {
+          await expectRejection(umbra.send(sender, dai.address, quantity, receiver.address), msg);
+        } else if (test.id === 'batchSend') {
+          await expectRejection(
+            umbra.batchSend(
+              sender,
+              [{ token: dai.address, amount: quantity, address: receiver!.publicKey }],
+              overrides
+            ),
+            msg
+          );
+        }
+      });
+
+      it('reverts if sender does not have enough ETH', async () => {
+        // ETH balance is checked by ethers when sending a transaction and therefore does not need to
+        // be tested here. If the user has insufficient balance it will throw with
+        // `insufficient funds for gas * price + value`
+      });
+
+      it('Send tokens, scan for them, withdraw them (direct withdraw)', async () => {
+        // SENDER
+        // Mint Dai to sender, and approve the Umbra contract to spend their DAI
+        await mintAndApproveDai(sender, sender.address, quantity.mul(5));
+
+        // Send funds with Umbra
+        let tx: ContractTransaction | null = null;
+        let stealthKeyPairs: KeyPair[] = [];
+        let usedReceivers: Wallet[] = [];
+
+        if (test.id === 'send') {
+          const result = await umbra.send(sender, dai.address, quantity, receiver!.publicKey, overrides);
+          tx = result.tx;
+          stealthKeyPairs = [result.stealthKeyPair];
+          usedReceivers = receivers.slice(0, 1);
```

Comment:
Hmm, I don't think this is correc? We send to `receiver` which uses the account at `receiverIndex`, but the account we save off in `usedReceivers` is from the `receivers` array which is a different index

Reply:
I pushed `receiver` into the receivers index way above so `receivers[0]` is `receiver`, but I see it's confusing. So,

1. I've removed the push code above, so `receivers` are a completely different set of wallets.
2. Made `usedReceivers = [receiver]`, so it's more clear.

Reply:
Code : https://github.com/ScopeLift/umbra-protocol/pull/426/commits/ddc768e7504ccb601877993210dbfa445f644277